### PR TITLE
Serialize fewer archetypes and tidy up casts

### PR DESCRIPTION
**Pull request by slavapestov**

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: lib/IRGen/DebugTypeInfo.h [hunk 1]
```diff
@@ -78,25 +78,29 @@ class DebugTypeInfo {
...
-    /// LValues, inout args, and Archetypes are implicitly indirect by
-    /// virtue of their DWARF type.
-    bool isImplicitlyIndirect() const {
-      return Type->isLValueType() || isArchetype() ||
-             (Type->getKind() == TypeKind::InOut);
-    }
+  /// LValues, inout args, and Archetypes are implicitly indirect by
+  /// virtue of their DWARF type.
+  //
+  // FIXME: Should this check if the lowered SILType is address only
```
Comment (slavapestov) [inline lib/IRGen/DebugTypeInfo.h]: @adrian-prantl What do you think? Should we be checking the SIL type lowering instead?

Also class-bound archetypes are not passed indirectly, so it's wrong in that case too.

File: lib/IRGen/IRGenSIL.cpp [hunk 2]
```diff
@@ -3278,9 +3278,13 @@ void IRGenSILFunction::visitDebugValueAddrInst(DebugValueAddrInst *i) {
   auto RealType = SILTy.getSwiftType();
   // Unwrap implicitly indirect types and types that are passed by
   // reference only at the SIL level and below.
+  //
+  // FIXME: Should this check if the lowered SILType is address only
```
Comment (slavapestov) [inline lib/IRGen/IRGenSIL.cpp]: @adrian-prantl Ditto

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: lib/IRGen/DebugTypeInfo.h (hunk 1 above)
Comment (adrian-prantl) [inline lib/IRGen/DebugTypeInfo.h]: Yes, please go ahead.

Comment (slavapestov) [inline lib/IRGen/DebugTypeInfo.h]: How can I construct a test case to verify my hypothesis? I don't want to start blindly changing this stuff without understanding what's going on :)

**Pull request merged**